### ISSUE TITLE
Dependency issues

### ISSUE THREAD

**Issue by tischi** [BIOP/bigdataviewer-bioformats#24]
@NicoKiaru 
Which version of `imglib2-cache` do we support?
Also for 
`<imglib2-cache.version>1.0.0-beta-15</imglib2-cache.version>`
which is the one that is in parent pom `29.2.1` 
 I get this error:
```
Exception in thread "main" java.lang.NoClassDefFoundError: net/imglib2/cache/img/DiskCachedCellImgOptions$CacheType
	at ch.epfl.biop.bdv.bioformats.bioformatssource.BioFormatsBdvOpener.getConcreteSource(BioFormatsBdvOpener.java:415
```

**Comment by NicoKiaru** [comment]
We can cut a release with the same dependencies as the ones of bdv-playground. This means that there's a few modifications to be done in this repo. Essentially this is what I reverted to make it compatible with BPD2.

Maybe it would be nice to cut a release right now (v0.2.11) which could still be used by BDP2. Then the next version (v0.2.12+) you will need to update BDP2 ?

**Comment by tischi** [comment]
Sounds good! I trust you, you have the better overview.
For BDP2 I can use relatively old versions.
For the project I am working on right now I need the boundingBoxCulling to work, because I want to show many small 2D sources.

**Comment by tischi** [comment]
@NicoKiaru 

Coming back to this, I'd need to ship update sites for those two guys:
- [bigdataprocessor2](https://github.com/bigdataprocessor/bigdataprocessor2), depending on
    - bdv-bioformats
    - [imagej-utils](https://github.com/tischi/imagej-utils)
- [segmentation-annotator](https://github.com/tischi/segmentation-annotator), depending on
    - bdv-bioformats
    - bdv-playground
    - bdv-core (with bounding box culling)
    - imagej-utils

Not sure whether imagej-utils will be an issue, and if so, I have to fix it. FYI, imagej-utils a core library of mine on which all my update sites depend. My current solution is to ship the imagej-utils.jar with all my update sites and try (pray) that everything is working with the latest version and also try to stay backward compatible.

Regarding all of this, I was wondering about the following things:

- [ ] Encourage BigStitcher to work with bdv-core 10.* ?
    - It should be actually an advantage for it, because it also should benefit from the bounding box culling. However, it may be some work for them since they may have to replace their own solution for this?!
- [ ] Should bdv-bioformats have its own update site? And all the update sites that depend on it (including bdv-playground) would mention in the installation that it is required to also enable it? 
    - Alternative is that all update sites ship their own bdv-bioformats.jar, but this may lead to incompatibilities.
    - I think it would deserve an own update site since it is very cool in its own right! You could even add a minimal Command to it that allows to run it.
- [ ] Solve the issue with imglib2?!
    - Did I get it right that the version of imglib2 that is currently given in parent-pom 29.2.1 is not working for us?
- [ ] Bdv-playground UI: If we also use bdv-playground as a core library (that's what I am doing right now), then I wonder whether its UI could be a bit more modest in terms of its appearance in Fiji. Could we just have the menus within the Plugins menu tree? An alternative could be a solution as bdp2 and bigstitcher employ it: when you run the plugin a new window pops up with all the functions. For bdv-playground the sources tree may be a natural place to add a menu bar to. I'll make an issue: https://github.com/bigdataviewer/bigdataviewer-playground/issues/163
- [ ] Anything else to consider?

**Comment by NicoKiaru** [comment]
> 
> 
> @NicoKiaru
> 
> Coming back to this, I'd need to ship update sites for those two guys:
> 
>     * [bigdataprocessor2](https://github.com/bigdataprocessor/bigdataprocessor2), depending on
>       
>       * bdv-bioformats
>       * [imagej-utils](https://github.com/tischi/imagej-utils)
> 
>     * [segmentation-annotator](https://github.com/tischi/segmentation-annotator), depending on
>       
>       * bdv-bioformats
>       * bdv-playground
>       * bdv-core (with bounding box culling)
>       * imagej-utils
> 
> 
> Not sure whether imagej-utils will be an issue, and if so, I have to fix it. FYI, imagej-utils a core library of mine on which all my update sites depend. My current solution is to ship the imagej-utils.jar with all my update sites and try (pray) that everything is working with the latest version and also try to stay backward compatible.
> 
> Regarding all of this, I was wondering about the following things:
> 
>     * [ ]  Encourage BigStitcher to work with bdv-core 10.* ?
>       
>       * It should be actually an advantage for it, because it also should benefit from the bounding box culling. However, it may be some work for them since they may have to replace their own solution for this?!

I have some PR opened on bigstitcher (https://github.com/PreibischLab/BigStitcher/pull/79) and multiview reconstruction (https://github.com/PreibischLab/multiview-reconstruction/pull/30). But I did not solve everything that's why they are draft. It could be nice to ping them again, but I'm not sure they will do anything before I2K ? 

> 
>     * [ ]  Should bdv-bioformats have its own update site? And all the update sites that depend on it (including bdv-playground) would mention in the installation that it is required to also enable it?
>       

I think it's ok to just declare the dependency in pom because they are accessible through maven scijava. And upload it on each update site. Personnally I find that having a compulsory update site dependency is a pain because it's hard to control and understand what's wrong for end users - for instance in macro scripting.  

>       * Alternative is that all update sites ship their own bdv-bioformats.jar, but this may lead to incompatibilities.

Maybe one option could be to try to contribute it in the parent pom scijava ? Regarding incompatibilities, I'm not sure we can be safe at the moment. Clearly I was aiming for a later release of my stuff and I updated the dependencies as soon as they were fixing bugs and introducing new useful features.

So far I'm stuck with:
- BigWarp v6.0.3-SNAPSHOT release (so not a reproducible release)
- BdvCore v10.0.2
- BdvVisTools also version above the parent pom
- Some fix in imglib2 cache to fix concurrency issue.

The way I'm dealing with all of this is uploading all these dependencies in the update site and shadowing the older version. The cost is : it breaks other update sites.

I'm ok with that for now. And the cost of keeping compatibility with the current dependencies of 29.2.1 as of today is too high for me.

We could ask for a timeline in gitter when the dependencies we required will make their way to the default fiji.

>       * I think it would deserve an own update site since it is very cool in its own right! You could even add a minimal Command to it that allows to run it.
> 

At the moment I think it's ok to ship the dependencies with each update site. But feel free to make an autonomous update site if you wish

>     * [ ]  Solve the issue with imglib2?!
>       
>       * Did I get it right that the version of imglib2 that is currently given in parent-pom 29.2.1 is not working for us?
> 

I think the issue may be with bdv-vistools and imglib2-cache. Not sure.

>     * [ ]  Bdv-playground UI: If we also use bdv-playground as a core library (that's what I am doing right now), then I wonder whether its UI could be a bit more modest in terms of its appearance in Fiji. Could we just have the menus within the Plugins menu tree?

Yes the commands need to disappear from the top level window. 

 An alternative could be a solution as bdp2 and bigstitcher employ it: when you run the plugin a new window pops up with all the functions. For bdv-playground the sources tree may be a natural place to add a menu bar to. I'll make an issue: [bigdataviewer/bigdataviewer-playground#163](https://github.com/bigdataviewer/bigdataviewer-playground/issues/163)
> 
>     * [ ]  Anything else to consider?

I'm a bit lost regarding all your projects ;-), and whether you want to make them accessible right now with the default fiji or not. My policy for now is :  I still need a stable release for BigWarp, and I'll try to stop increasing the dependencies in order to have something that works in the default fiji for let's say beginning of 2021. By that time, I ship each independent plugin that brings their needed updated dependencies and which can break other update sites.

**Comment by tischi** [comment]
> By that time, I ship each independent plugin that brings their needed updated dependencies and which can break other update sites.

OK, then I'll do that, too, for now. 

If we give courses ( like the big image data webinars ) we have to coordinate with the bigwarp and bigstitcher people to make sure things are running for the participants.

Currently bdv-playground does _not_ have bdv-bioformats as a dependency in the pom. Is that on purpose?

**Comment by NicoKiaru** [comment]
> If we give courses ( like the big image data webinars ) we have to coordinate with the bigwarp and bigstitcher people to make sure things are running for the participants.

BigWarp should work with the 6.0.3-SNAPSHOT dependency. BigStitcher does not work unless you take my branch for multiview-reconstruction and bigstitcher (quite annoying)

> Currently bdv-playground does not have bdv-bioformats as a dependency in the pom. Is that on purpose?

Yes. I think it makes sense. And it was needed for BDP2 to split these repos and have bdv-bioformats compatible with pom scijava 29.2.1

**Comment by tischi** [comment]
But how do you then use bdv-bioformats from within the playground? Or don't you?

**Comment by NicoKiaru** [comment]
I don't when I develop bdv-playground in intellij. When I have a repo which needs both, then I add bdv-playground and bdv-bioformats as dependencies, like in [bigdataviewer-biop-tools](https://github.com/BIOP/bigdataviewer-biop-tools/blob/master/pom.xml). Same for update sites : I upload both if I need both. Makes sense ?

**Comment by NicoKiaru** [comment]
Ah but because of the current compatibility issue, I keep bdv-bioformats 0.2.9 for the moment when I need both.

**Comment by tischi** [comment]
Hi @NicoKiaru ,

Would you be able to provide me with a list of compatible `version numbers of
- bdv-playground
- bdc-bioformats
- imglib2-cache

for the two scenarios:

1. BigDataProcessor2 (should be compatible with current Fiji)
2. Latest (contains the boundingBoxCulling, will break stuff in Fiji)

**Comment by NicoKiaru** [comment]
So for question 1 : 
Ouch, for bdv-playground you would have to go back in time a lot... And the version was not in scijava maven. This means you have to look at the commits far away in the past. + the dependencies may have been snapshot at the time.
-> I'm not able.
Question 2:
* <imglib2-cache.version>1.0.0-beta-15</imglib2-cache.version>
* <bigdataviewer-playground.version>0.2.7-SNAPSHOT</bigdataviewer-playground.version>
* <bigdataviewer-bioformats.version>0.2.9</bigdataviewer-bioformats.version>

**Comment by tischi** [comment]
Question1:
I think there we in fact need only bdv-bioformats, is there a version that I can take?

**Comment by NicoKiaru** [comment]
1:
<bigdataviewer-bioformats.version>0.2.10</bigdataviewer-bioformats.version>
or the current snapshot works

**Comment by tischi** [comment]
Question 2:
Can I also take a newer version of bdv-bf, which contains our fixes  for the source and channel names?

**Comment by tischi** [comment]
Ok, Looks like I cannot: `java.lang.ClassNotFoundException: net.imglib2.cache.img.DiskCachedCellImgOptions$CacheType`

**Comment by NicoKiaru** [comment]
2:

Not yet. I can do it. Then 0.2.11 will remain ok for BDP2, and 0.2.12-SNAPSHOT will be updated. But for now, only 0.2.9 is compatible with bdv-playground

**Comment by NicoKiaru** [comment]
bdv-bf v0.2.11 pushed (compatible with BDP2)

**Comment by tischi** [comment]
confused: which version of bdv-bf can I now use for scenario 2?

**Comment by NicoKiaru** [comment]
> 
> 
> confused: which version of bdv-bf can I now use for scenario 2?

Right now : 0.2.9 only.

0.2.9 breaks stuff in FIJI, and is compatible with bdv-pg
0.2.10 do not break FIJI, is compatible with bdp2 (not with bdv-pg)
0.2.11 do not break FIJI, is compatible with bdp2 (pretty sure) (not with bdv-pg)
0.2.12+ will break FIJI, and will be compatible with bdv-pg

**Comment by NicoKiaru** [comment]
Hmmmmm I'm kind of lost, it's not really behaving as I expected.

Try 0.2.9, 0.2.10, 0.2.11 and 0.2.12-SNAPSHOT... And see what happens...

**Comment by tischi** [comment]
@NicoKiaru 

For me those versions work now together (this would be my scenario 2, i.e. LATEST):
```
		<bigdataviewer-bioformats.version>0.2.11-SNAPSHOT</bigdataviewer-bioformats.version>
		<bigdataviewer-playground.version>0.2.7-SNAPSHOT</bigdataviewer-playground.version>
		<imglib2-cache.version>1.0.0-beta-15</imglib2-cache.version>
```

However, I was adding to `bigdataviewer-bioformats` the `imglib2-cache` dependency and re-building it locally with this dependency. My current theory is that your latest builds of `bigdataviewer-bioformats` did not include this version of `imglib2-cache`.  What I found weird is that in your pom my IntelliJ rendered the `<imglib2-cache.version>1.0.0-beta-15</imglib2-cache.version>` gray, indicating that is was ignored. In fact, I saw that it was fetching the version from the parent pom instead. That's why I added it explicitly as shown below. 

I got confused several times regarding the precedence of versions if they are specified at several places. I think the safest is to add them explicitly like below, but it is not very nice.

```
		<dependency>
			<groupId>net.imglib2</groupId>
			<artifactId>imglib2-cache</artifactId>
			<version>1.0.0-beta-15</version>
		</dependency>
```
